iocore/net/SSLUtils.cc

```diff
@@ -412,12 +424,48 @@ ssl_client_hello_callback(SSL *s, int *al, void *arg)
 }
 #endif
 
+#ifdef OPENSSL_IS_BORINGSSL
+static ssl_select_cert_result_t
+ssl_client_hello_callback(const SSL_CLIENT_HELLO *client_hello)
```

Comment:
Maybe you want a different name for the BORINGSSL version of this function, which is really a utility function for the callback instead of the proper callback.  Took me a few minutes looking around to resolve in my head that the boring version of the function and openssl version with the same name weren't going to conflict with each other.

Reply:
I don't think this function should have a different name, although it's a shame that we have two very similar functions.

The lambda expression in `_set_handshake_callbacks` mimics OpenSSL's behavior, and it enables reusing two other callback functions as well. In that sense I think having the same name make sense.

Having this in a `#elif` clause may ease the confusion.
```cpp
#if TS_USE_HELLO_CB
static int
ssl_client_hello_callback(SSL *s, int *al, void *arg)
{
}
#elif defined(OPENSSL_IS_BORINGSSL)
static ssl_select_cert_result_t
ssl_client_hello_callback(const SSL_CLIENT_HELLO *client_hello)
{
}
#endif
```

And we may also be able to have exactly the same function signature by passing `SSL_CLIENT_HELLO` to `arg`.

Reply:
Doing the #elif here makes sense.

The callback return values are different types (and different values for equivalent states):

```
enum ssl_select_cert_result_t BORINGSSL_ENUM_INT {
  // ssl_select_cert_success indicates that the certificate selection was
  // successful.
  ssl_select_cert_success = 1,
  // ssl_select_cert_retry indicates that the operation could not be
  // immediately completed and must be reattempted at a later point.
  ssl_select_cert_retry = 0,
  // ssl_select_cert_error indicates that a fatal error occured and the
  // handshake should be terminated.
  ssl_select_cert_error = -1,
};
```
vs openssl:
```
# define SSL_CLIENT_HELLO_SUCCESS 1
# define SSL_CLIENT_HELLO_ERROR   0
# define SSL_CLIENT_HELLO_RETRY   (-1)
```

Reply:
I'm fine with the current code, but I guess we could have aliases for the type and also the values.
https://github.com/apache/trafficserver/blob/870bbdb18eb267e4fd0ebebf3d667ac0dd1d5479/iocore/net/P_SSLUtils.h#L45-L49